### PR TITLE
feat(planner): Implement hash inner join

### PR DESCRIPTION
**Pull request by leiysky**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Initial version of hash join.

Only support inner join with `ON` clause, for example:
```sql
select * from t inner join t1 on t.a = t1.a
```

## Changelog

- New Feature

## Related Issues

Fixes #issue

### CODE REVIEW

**Comment by Xuanwo** [comment]
Join the world!

File: query/src/interpreters/interpreter_select_v2.rs [hunk 1]
```diff
@@ -52,9 +53,17 @@ impl Interpreter for SelectInterpreterV2 {
...
-        let executor = PipelinePullingExecutor::try_create(async_runtime, pipeline)?;
+
+        // Spawn sub-pipelines
+        for pipeline in pipelines {
+            let executor = PipelineExecutor::create(async_runtime.clone(), pipeline)?;
+            executor.execute()?;
+        }
+
+        // Spawn root pipeline
+        let executor = PipelinePullingExecutor::try_create(async_runtime, root_pipeline)?;
```
[the remark this answers is not in the document]
Comment (zhang2014) [reply query/src/interpreters/interpreter_select_v2.rs:66]: spawn sub-tasks is ok. I'll work on this task.

We need add NewPipe type(zip nested pipeline last pipe with outer pipeline previous pipe):
```rust
NewPipe::NestedPipeline { 
    nested_pipeline: NewPipeline,
    processors: Vec<ProcessorPtr>, 
    outputs_port: Vec<Arc<OutputPort>>, 
    inputs_port: Vec<(Arc<InputPort>, Arc<InputPort>)>, 

}
```

File: query/src/sql/exec/mod.rs [hunk 2]
```diff
@@ -307,40 +359,109 @@ impl PipelineBuilder {
...
-                    AggregatorTransformParams::try_create(
-                        transform_input_port,
-                        transform_output_port,
-                        &final_aggr_params,
-                    )?,
-                    self.ctx.clone(),
-                )
-            })?;
+
+        pipeline.resize(1)?;
```
Comment (xudong963) [inline query/src/sql/exec/mod.rs:382]: You added in the PR 🤣

File: query/src/pipelines/new/processors/transforms/hash_join/hash_table.rs [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+                ExpressionEvaluator::eval(self.ctx.try_get_function_context()?, expr, &input)
+            })
+            .collect::<Result<Vec<ColumnRef>>>()?;
+
+        let hash_values = self.hash(&build_keys, input.num_rows())?;
+
+        let mut null_bitmap: Option<Bitmap> = None;
+        for column in build_keys.iter() {
+            let (_, bitmap) = column.validity();
+            null_bitmap = combine_validities(null_bitmap.as_ref(), bitmap);
```
Comment (sundy-li) [inline query/src/pipelines/new/processors/transforms/hash_join/hash_table.rs:203]: null_bitmap is useless now ?

Comment (leiysky) [reply query/src/pipelines/new/processors/transforms/hash_join/hash_table.rs:203]: Yes, I forgot to remove it.

Comment (leiysky) [reply query/src/pipelines/new/processors/transforms/hash_join/hash_table.rs]: Fixed.

File: query/src/sql/optimizer/rule/rule_implement_hash_join.rs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    fn id(&self) -> RuleID {
+        self.id
+    }
+
+    fn apply(&self, expression: &SExpr, state: &mut TransformState) -> Result<()> {
+        let plan = expression.plan();
+        let logical_inner_join: LogicalInnerJoin = plan.try_into()?;
+
+        let result = SExpr::create(
+            PhysicalHashJoin {
```
Comment (xudong963) [inline query/src/sql/optimizer/rule/rule_implement_hash_join.rs:67]: In the future, when our statistics module matures, it seems that the build side can be determined based on statistical information

File: query/src/pipelines/new/processors/transforms/hash_join/hash_table.rs [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+
+        result
+    }
+
+    fn hash(&self, columns: &[ColumnRef], row_count: usize) -> Result<HashVector> {
+        let columns_vec: Vec<ColumnRef> = columns.to_vec();
+        let columns: Vec<&ColumnRef> = columns_vec.iter().collect();
+        Ok(match &self.hash_method_kind {
+            HashMethodKind::Serializer(method) => method
+                .build_keys(&columns, row_count)?
```
Comment (sundy-li) [inline query/src/pipelines/new/processors/transforms/hash_join/hash_table.rs:120]: I did not find any reason to use `hash_method_kind`.
Evaluate [`xxhash64`](https://github.com/datafuselabs/databend/blob/752ecb112c68d22d65298dd309aab41a9ceaa7bd/common/functions/src/scalars/hashes/hash.rs) may be more simple way to do it.

File: query/src/sql/exec/data_schema_builder.rs [hunk 6]
```diff
@@ -134,4 +135,17 @@ impl<'a> DataSchemaBuilder<'a> {
...
+        // TODO: NATURAL JOIN and USING
+        let mut fields = Vec::with_capacity(left.num_fields() + right.num_fields());
+        for field in left.fields().iter() {
+            fields.push(field.clone());
+        }
+        for field in right.fields().iter() {
+            fields.push(field.clone());
+        }
+
+        DataSchemaRefExt::create(fields)
```
Comment (xudong963) [inline query/src/sql/exec/data_schema_builder.rs:149]: 👍, other build_xxx methods can also use the API.

File: query/src/pipelines/new/processors/transforms/hash_join/hash_table.rs (hunk 5 above)
Comment (leiysky) [reply query/src/pipelines/new/processors/transforms/hash_join/hash_table.rs:120]: I was trying to reuse the `HashMethod`, but it seems the need of hashing is different from `Aggregate`.

It only needs a hashed `u64` value here. Maybe I can just compute hash for each column and combine them together?

File: query/src/sql/optimizer/rule/rule_implement_hash_join.rs (hunk 4 above)
Comment (leiysky) [reply query/src/sql/optimizer/rule/rule_implement_hash_join.rs:67]: Yes. We will always choose right side as build side. The join commutativity rule can help us enumurate the candidates.

File: query/src/pipelines/new/processors/transforms/hash_join/hash_table.rs (hunk 5 above)
Comment (sundy-li) [reply query/src/pipelines/new/processors/transforms/hash_join/hash_table.rs:120]: `HashMethod ` is optimized in the case of group aggregation, it produces unique keys and leads to more memory.

Now you only need a hashed u64 value, so `HashMethod` may be useless unless you store your state using HashTable.

Comment (sundy-li) [reply query/src/pipelines/new/processors/transforms/hash_join/hash_table.rs:120]: I got your point, now hash function may not work for multi-columns.  Let's keep it as now and refactor it in the future.

**Comment by BohuTANG** [comment]
Oops, Conflicting files :\

**Comment by leiysky** [comment]
> Oops, Conflicting files :\

I'm fixing that.

File: query/src/pipelines/new/processors/transforms/hash_join/hash_table.rs (hunk 5 above)
Comment (leiysky) [reply query/src/pipelines/new/processors/transforms/hash_join/hash_table.rs]: @sundy-li I have remove the `HashMethod` and use hash `Function` instead. PTAL

File: tests/suites/0_stateless/20+_others/20_0001_planner_v2.sql [hunk 7]
```diff
@@ -43,4 +43,22 @@ select count(null) from numbers(1000);
...
+create table t(a int);
+insert into t values(1),(2),(3);
+create table t1(b float);
+insert into t1 values(1.0),(2.0),(3.0);
+create table t2(c uint32 null);
+insert into t2 values(1),(2),(null);
+
+select * from t inner join t1 on cast(t.a as float) = t1.b;
+select * from t inner join t2 on t.a = t2.c;
+select * from t inner join t2 on t.a = t2.c + 1;
```
Comment (sundy-li) [inline tests/suites/0_stateless/20+_others/20_0001_planner_v2.sql:57]: I did not get it, why does this query have empty results?

Comment (leiysky) [reply tests/suites/0_stateless/20+_others/20_0001_planner_v2.sql:57]: It looks like a bug, this query should output:
```
a c
----
2 1
3 2
```

Just as the following query.

I'll fix it.

Comment (BohuTANG) [reply tests/suites/0_stateless/20+_others/20_0001_planner_v2.sql:57]: `select null+1;` will panic in v2:
```
022-05-07T14:07:37.182626Z ERROR common_tracing::panic_hook: panicked at 'called `Option::unwrap()` on a `None` value', query/src/sql/planner/binder/project.rs:44:59 backtrace=Backtrace [{ fn: "common_tracing::panic_hook::set_panic_hook::{{closure}}", file: "./common/tracing/src/panic_hook.rs", line: 25 }, { fn: "std::panicking::rust_panic_with_hook", 
```

**Pull request merged**